File: google-ai-mode-scraper/browser/python/google_ai_mode.py
"""GoogleAiMode scraper using the official Scrapeless Python SDK + Playwright over CDP.

Google's AI Mode is the streaming AI overlay reachable at
``https://www.google.com/search?q=<query>&udm=50``. The scraper drives the
SERP, waits for the AI panel to stream in, and extracts the answer text
plus cited links.

Flow:
- ``client.browser.create()`` mints a cloud session on Scrapeless's Scraping Browser.
- Playwright connects over CDP, navigates to the SERP with ``udm=50``.
- Wait for the AI panel container, then settle for streamed content.
- Parsel parses the rendered HTML into a single ``AiResponse`` dataclass.
"""

from __future__ import annotations

import asyncio
import os
import re
from dataclasses import asdict, dataclass, field
from typing import Any
from urllib.parse import quote_plus, urlparse

from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 240
AI_READY_SELECTOR = "div[data-subtree='aimc'], div[jsname][data-md], div[role='region']"

# ---------------------------------------------------------------------------
# Schemas
# ---------------------------------------------------------------------------


@dataclass
class Citation:
    title: str
    url: str
    source: str


@dataclass
class AiLink:
    url: str
    text: str


@dataclass
class AiResponse:
    query: str
    url: str
    response_text: str
    citations: list[Citation] = field(default_factory=list)
    links: list[AiLink] = field(default_factory=list)


# ---------------------------------------------------------------------------
# Scrapeless plumbing
# ---------------------------------------------------------------------------


def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()


async def _settle_for_streamed_content(page, *, idle_ms: int = 3000, ceiling_ms: int = 30000) -> None:
    """Wait for the AI panel to stop changing. Google paints the answer
    incrementally; we bail once ``document.body.innerText.length`` is stable
    for ``idle_ms`` or after a hard ceiling.
    """
    loop = asyncio.get_event_loop()
    start = loop.time()
    last_len = 0
    last_change = loop.time()
    while (loop.time() - start) * 1000 < ceiling_ms:
        await asyncio.sleep(0.5)
        try:
            length = await page.evaluate(
                "() => document.body && document.body.innerText ? document.body.innerText.length : 0"
            )
        except Exception:
            length = 0
        if length != last_len:
            last_len = length
            last_change = loop.time()
        elif (loop.time() - last_change) * 1000 >= idle_ms:
            return


async def _warmup_google(page) -> None:
    """Land on the bare homepage first so Google's NID/consent cookies are
    minted on a less-defended endpoint before hitting the AI Mode SERP.
    """
    try:
        await page.goto("https://www.google.com/", wait_until="domcontentloaded", timeout=30000)
        for selector in ("button[aria-label*='Accept']", "#L2AGLb"):
            try:
                await page.click(selector, timeout=1500)
            except Exception:
                pass
        await asyncio.sleep(1.5)
    except Exception:
        pass


def _is_captcha(page) -> bool:
    u = page.url or ""
    return "/sorry/" in u or "captcha" in u


async def _fetch_ai_mode_html(
    query: str, *, proxy_country: str = DEFAULT_PROXY_COUNTRY, retries: int = 3
) -> tuple[str, str]:
    url = f"https://www.google.com/search?q={quote_plus(query)}&udm=50&hl=en"
    last_error: Exception | None = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
        )
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(session.browser_ws_endpoint)
            try:
                page = await browser.new_page(viewport={"width": 1366, "height": 900})
                await _warmup_google(page)
                await page.goto(url, wait_until="domcontentloaded", timeout=45000)
                if _is_captcha(page):
                    last_error = RuntimeError("captcha interstitial")
                else:
                    try:
                        await page.wait_for_selector(AI_READY_SELECTOR, timeout=20000)
                    except Exception as e:
                        logger.warning("ai panel selector not seen (continuing): {}", e)
                    await _settle_for_streamed_content(page)
                    if _is_captcha(page):
                        last_error = RuntimeError("captcha interstitial")
                    else:
                        html = await page.content()
                        if html:
                            return html, page.url
                        last_error = RuntimeError("empty HTML")
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("attempt {} failed: {}", attempt + 1, e)
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
        if attempt < retries:
            await asyncio.sleep(4 + attempt * 3)
    raise RuntimeError(f"giving up after {retries + 1} attempts: {last_error}")


# ---------------------------------------------------------------------------
# Scrape functions
# ---------------------------------------------------------------------------


async def scrape_ai_response(query: str) -> AiResponse:
    html, final_url = await _fetch_ai_mode_html(query)
    return parse_ai_response(html, query, final_url)


# ---------------------------------------------------------------------------
# Parser
# ---------------------------------------------------------------------------


_WS_RE = re.compile(r"\s+")


def _normalise(text: str) -> str:
    return _WS_RE.sub(" ", text or "").strip()


def _domain_of(url: str) -> str:
    try:
        host = (urlparse(url).hostname or "").lower()
        if host.startswith("www."):
            host = host[4:]
        return host
    except Exception:
        return ""


def parse_ai_response(html: str, query: str, final_url: str) -> AiResponse:
    sel = Selector(text=html)

    candidates = [
        "div[data-subtree='aimc']",
        "div[jsname][data-md]",
        "div[role='region'][aria-label]",
        "#main",
    ]
    panel = None
    for css in candidates:
        for node in sel.css(css):
            text_content = "".join(node.css("*::text").getall()).strip()
            if len(text_content) > 80:
                panel = node
                break
        if panel is not None:
            break
    if panel is None:
        panel = sel.css("body")

    raw_text = " ".join(panel.css("*::text").getall())
    response_text = _normalise(raw_text)

    citations: list[Citation] = []
    seen_cite: set[str] = set()
    links: list[AiLink] = []
    seen_link: set[str] = set()

    for a in panel.css("a[href^='http']"):
        href = a.attrib.get("href", "")
        if not href or "google.com/search" in href or "/url?" in href:
            continue
        text = _normalise("".join(a.css("*::text").getall()))
        aria = a.attrib.get("aria-label", "")
        has_favicon = bool(a.css("img"))
        source = _domain_of(href)

        if has_favicon or aria:
            key = f"{href}|{text}"
            if key not in seen_cite:
                seen_cite.add(key)
                citations.append(Citation(title=text or aria or source, url=href, source=source))

        if href not in seen_link:
            seen_link.add(href)
            links.append(AiLink(url=href, text=text or aria))

    return AiResponse(
        query=query,
        url=final_url or f"https://www.google.com/search?q={quote_plus(query)}&udm=50",
        response_text=response_text,
        citations=citations,
        links=links,
    )


def to_dict(obj: Any) -> Any:
    if hasattr(obj, "__dataclass_fields__"):
        return asdict(obj)
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj

File: google-ai-mode-scraper/browser/nodejs/google-ai-mode.mjs
// GoogleAiMode scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// Google's AI Mode is the streaming AI overlay reachable at
// `https://www.google.com/search?q=<query>&udm=50`. The scraper drives the
// SERP, waits for the AI panel to stream in, and extracts the answer text
// plus cited links.
//
// Flow:
// - `client.browser.create()` mints a cloud session (CDP WebSocket).
// - puppeteer-core connects, navigates to the SERP with `udm=50`.
// - Wait for the AI panel container, then settle for streamed content.
// - cheerio parses the rendered HTML into a single `AiResponse` object.

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 240;
const AI_READY_SELECTOR = "div[data-subtree='aimc'], div[jsname][data-md], div[role='region']";

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function warmupGoogle(page) {
  // Land on the bare homepage so Google's NID/consent cookies are minted on
  // a less-defended endpoint before the AI Mode SERP. The first hop often
  // lands on /sorry/ without this.
  try {
    await page.goto("https://www.google.com/", { waitUntil: "domcontentloaded", timeout: 30000 });
    try { await page.click("button[aria-label*='Accept']", { timeout: 1500 }); } catch (_) {}
    try { await page.click("#L2AGLb", { timeout: 1500 }); } catch (_) {}
    await new Promise((r) => setTimeout(r, 1500));
  } catch (_) {}
}

function isCaptchaPage(page) {
  const url = page.url() || "";
  return url.includes("/sorry/") || url.includes("captcha");
}

async function fetchAiModeHtml(query, { proxyCountry = DEFAULT_PROXY_COUNTRY, retries = 3 } = {}) {
  const url = `https://www.google.com/search?q=${encodeURIComponent(query)}&udm=50&hl=en`;
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const { browserWSEndpoint } = await client().browser.create({
      proxyCountry,
      sessionTTL: DEFAULT_SESSION_TTL,
    });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint });
      const page = await browser.newPage();
      await page.setViewport({ width: 1366, height: 900 });
      await warmupGoogle(page);
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 45000 });
      if (isCaptchaPage(page)) {
        lastError = new Error("captcha interstitial");
      } else {
        try { await page.waitForSelector(AI_READY_SELECTOR, { timeout: 20000 }); } catch (_) {}
        // Let the streamed AI answer settle. Google paints the AI panel
        // incrementally; bail once the DOM stops changing for 3 seconds or
        // after a hard ceiling.
        await settleForStreamedContent(page, { idleMs: 3000, ceilingMs: 30000 });
        if (isCaptchaPage(page)) {
          lastError = new Error("captcha interstitial");
        } else {
          const html = await page.content();
          if (html) return { html, finalUrl: page.url() };
          lastError = new Error("empty HTML");
        }
      }
    } catch (e) {
      lastError = e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
    if (attempt < retries) await new Promise((r) => setTimeout(r, 4000 + attempt * 3000));
  }
  throw new Error(`giving up after ${retries + 1} attempts: ${lastError?.message}`);
}

async function settleForStreamedContent(page, { idleMs, ceilingMs }) {
  const start = Date.now();
  let lastLen = 0;
  let lastChange = Date.now();
  while (Date.now() - start < ceilingMs) {
    await new Promise((r) => setTimeout(r, 500));
    let len = 0;
    try { len = await page.evaluate(() => document.body?.innerText?.length ?? 0); } catch (_) {}
    if (len !== lastLen) {
      lastLen = len;
      lastChange = Date.now();
    } else if (Date.now() - lastChange >= idleMs) {
      return;
    }
  }
}

// ---------------- scrape functions ----------------

export async function scrapeAiResponse(query) {
  const { html, finalUrl } = await fetchAiModeHtml(query);
  return parseAiResponse(html, query, finalUrl);
}

// ---------------- parser ----------------

export function parseAiResponse(html, query, finalUrl) {
  const $ = cheerio.load(html);

  // The AI panel renders inside one of a handful of containers. Try the
  // most specific selectors first, then fall back to the main content
  // column. We deliberately keep this defensive — Google rerolls AI Mode
  // markup frequently.
  const candidates = [
    "div[data-subtree='aimc']",
    "div[jsname][data-md]",
    "div[role='region'][aria-label]",
    "#main",
  ];
  let $panel = null;
  for (const sel of candidates) {
    const node = $(sel).filter((_, el) => $(el).text().trim().length > 80).first();
    if (node.length) { $panel = node; break; }
  }
  if (!$panel || !$panel.length) $panel = $("body");

  const responseText = normaliseText($panel.text());

  // Citations: chips/cards rendered with an outbound link + visible label.
  // Google decorates them with `aria-label` containing the source domain.
  const citations = [];
  const seenCite = new Set();
  $panel.find("a[href^='http']").each((_, a) => {
    const $a = $(a);
    const href = $a.attr("href") || "";
    if (!href || href.includes("google.com/search") || href.includes("/url?")) return;
    const aria = $a.attr("aria-label") || "";
    const text = normaliseText($a.text());
    // Citations show a domain badge — most have an `<img>` favicon descendant.
    const hasFavicon = $a.find("img").length > 0;
    if (!hasFavicon && !aria) return;
    const source = domainOf(href);
    const key = `${href}|${text}`;
    if (seenCite.has(key)) return;
    seenCite.add(key);
    citations.push({
      title: text || aria || source,
      url: href,
      source,
    });
  });

  // Generic outbound links (super-set of citations, deduped on href alone).
  const links = [];
  const seenLink = new Set();
  $panel.find("a[href^='http']").each((_, a) => {
    const $a = $(a);
    const href = $a.attr("href") || "";
    if (!href || href.includes("google.com/search") || href.includes("/url?")) return;
    if (seenLink.has(href)) return;
    seenLink.add(href);
    links.push({
      url: href,
      text: normaliseText($a.text()) || ($a.attr("aria-label") || ""),
    });
  });

  return {
    query,
    url: finalUrl || `https://www.google.com/search?q=${encodeURIComponent(query)}&udm=50`,
    response_text: responseText,
    citations,
    links,
  };
}

function normaliseText(s) {
  return String(s || "").replace(/\s+/g, " ").trim();
}

function domainOf(url) {
  try {
    let host = new URL(url).hostname || "";
    if (host.startsWith("www.")) host = host.slice(4);
    return host;
  } catch {
    return "";
  }
}
